components/push/src/communications/rate_limiter.rs

```diff
@@ -0,0 +1,163 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::storage::{Storage, Store};
+use std::time::{SystemTime, UNIX_EPOCH};
+
+// DB persisted rate limiter.
+// Implementation notes: This saves the timestamp of our latest call and the number of times we have
+// called `Self::check` within the `Self::periodic_interval` interval of time.
+pub struct PersistedRateLimiter {
+    op_name: String,
+    periodic_interval: u64, // In seconds.
+    max_requests_in_interval: u16,
+}
+
+impl PersistedRateLimiter {
+    pub fn new(op_name: &str, periodic_interval: u64, max_requests_in_interval: u16) -> Self {
+        Self {
+            op_name: op_name.to_owned(),
+            periodic_interval,
+            max_requests_in_interval,
+        }
+    }
+
+    pub fn check(&self, store: &Store) -> bool {
+        let (mut timestamp, mut count) = self.get_counters(store);
+
+        let now = now_secs();
+        if (now - timestamp) >= self.periodic_interval {
+            log::info!(
+                "Resetting. now({}) - {} < {} for {}.",
+                now,
+                timestamp,
+                self.periodic_interval,
+                &self.op_name
+            );
+            count = 0;
+            timestamp = now;
+        } else {
+            log::info!(
+                "No need to reset inner timestamp and count for {}.",
+                &self.op_name
+            )
+        }
+
+        count += 1;
+        self.persist_counters(store, timestamp, count);
+
+        // within interval counter
+        if count > self.max_requests_in_interval {
+            log::info!(
+                "Not allowed: innerCount($innerCount) > $MAX_REQUEST_IN_INTERVAL for {}.",
+                &self.op_name
+            );
+            return false;
+        }
+
+        log::info!("Allowed to pass through for {}!", &self.op_name);
+
+        true
+    }
+
+    fn db_meta_keys(&self) -> (String, String) {
+        (
+            format!("ratelimit_{}_timestamp", &self.op_name),
+            format!("ratelimit_{}_count", &self.op_name),
+        )
+    }
+
+    fn get_counters(&self, store: &Store) -> (u64, u16) {
+        let (timestamp_key, count_key) = self.db_meta_keys();
+        (
+            self.get_meta_integer(store, &timestamp_key),
+            self.get_meta_integer(store, &count_key) as u16, // We know count_key is stored as a `u16`, so there's no harm doing a cast here.
```

Comment:
FWIW I think the fully-generic version would have worked if you made it generic over anything implementing `FromStr` rather than over integers, e.g.:

* https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a5512d3f9fff7331e54db7c2f5897845

But I'm fine the cast here in practice.

Reply:
Very nice!